src/Microsoft.SqlTools.ServiceLayer/Connection/ListDatabaseRequestHandler.cs

```diff
@@ -64,39 +67,75 @@ abstract class ListDatabaseRequestHandler<T> : IListDatabaseRequestHandler
 
         public ListDatabasesResponse HandleRequest(ISqlConnectionFactory connectionFactory, ConnectionInfo connectionInfo)
         {
+            ListDatabasesResponse response = new ListDatabasesResponse();
             ConnectionDetails connectionDetails = connectionInfo.ConnectionDetails.Clone();
+            // Running query against sys.databases view will only return a subset of databases the current login/user might have access to, we need to
+            // query the master database to get the full database list, but for users without master db access, we have to query the
+            // original database as a fallback.
+            var databasesToTry = new List<string>() { CommonConstants.MasterDatabaseName };
```

Comment:
For Azure SQL DB it's way more common for users not to have any access to databases other than the specific one they connected to. Querying `master` is likely to be a waste of time. Even if they do have limited access to master, a successful query will likely lead to the list being populated with potentially a huge number of databases, many of which the user can't access. Can't we just fail gracefully if the attempt to access master fails?

The other option is if the connection string specifies a non-master database and the database is Azure DB, we just remove master from the list for switching. I consider it somewhat unfortunate that `sys.databases` in Azure DB includes `master` when querying a user database.

Reply:
Agreed, it is all because of SQL server includes master and tempdb (on prem only) when querying a userdb with contained user. I was also thinking about only return the current database name for this scenario.

But we can't simply remove master from the list when connecting to a user database directly, see the issue I mentioned in the PR description: https://github.com/microsoft/azuredatastudio/issues/10816. we have to query the master first and then fallback.